pyro/distributions/permutation_flow.py

```diff
@@ -0,0 +1,78 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.distributions.transforms import Transform
+from torch.distributions import constraints
+
+from pyro.distributions.util import copy_docs_from
+
+
+@copy_docs_from(Transform)
+class PermutationFlow(Transform):
+    """
+    A normalizing flow that reorders the input dimensions, that is, multiplies the input by a permutation matrix.
+    This is useful in between IAF transforms to increase the flexibility of the resulting distribution and
+    stabilize learning. Whilst not being an autoregressive flow, the log absolute determinate of the Jacobian is
+    easily calculable as 0. Note that reordering the input dimension between two layers of IAF is not equivalent
+    to reordering the dimension inside the MADE networks that those IAFs use; using a PermutationFlow results in a
+    distribution with more flexibility.
+
+    Example usage:
+
+    >>> from pyro.nn import AutoRegressiveNN
+    >>> from pyro.distributions import InverseAutoregressiveFlow, PermutationFlow
+    >>> base_dist = dist.Normal(torch.zeros(10), torch.ones(10))
+    >>> iaf1 = InverseAutoregressiveFlow(AutoRegressiveNN(10, [40]))
+    >>> ff = PermutationFlow(torch.randperm(10, dtype=torch.long))
+    >>> iaf2 = InverseAutoregressiveFlow(AutoRegressiveNN(10, [40]))
+    >>> iaf_dist = dist.TransformedDistribution(base_dist, [iaf1, ff, iaf2])
+    >>> iaf_dist.sample()  # doctest: +SKIP
+        tensor([-0.4071, -0.5030,  0.7924, -0.2366, -0.2387, -0.1417,  0.0868,
+                0.1389, -0.4629,  0.0986])
+
+    :param permutation: a permutation ordering that is applied to the inputs.
+    :type permutation: torch.LongTensor
+
+    """
+
+    codomain = constraints.real
+
+    def __init__(self, permutation):
+        super(PermutationFlow, self).__init__()
+
+        self.permutation = permutation
+
+        # Calculate the inverse permutation order
+        self.inv_permutation = torch.empty(permutation.shape, dtype=torch.long)
```

Comment:
I think you'll need to add `device=permutation.device` to the following two lines.

Reply:
Done